### PR TITLE
V2: Refactor `MLP` class and delegate encoding to `Predictor`

### PR DESCRIPTION
**Pull request by shihchengli**
The PR #757 has been reverted in PR #759. This PR will be used to continue the work of the original PR.

### CODE REVIEW

**Review by KnathanM — APPROVED** [review]

**Comment by KnathanM** [comment]
Both @davidegraff and I have approved, so this can be merged in. I left one suggested change that is stylistic. Up to @shihchengli if he wants to add it.

File: chemprop/cli/train.py [hunk 1]
```diff
@@ -686,8 +686,9 @@ def build_model(args, train_dset: MolGraphDataset | MulticomponentDataset) -> MP
         )
         model.bn.apply(lambda module: module.requires_grad_(False))
         for idx in range(args.frzn_ffn_layers):
-            model.predictor.ffn[idx * 3].requires_grad_(False)
-            setattr(model.predictor.ffn[idx * 3 + 2], "p", 0.0)
+            suffix = 0 if idx == 0 else 2
+            model.predictor.ffn[idx][suffix].requires_grad_(False)
+            setattr(model.predictor.ffn[idx + 1][1], "p", 0.0)
```
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:690]: I'm not sure if it's correct to freeze the subsequent dropout layer. I'm concerned that it might lead to overfitting for the task. For instance, consider a MLP with two blocks:
0 0 Linear
1 0 ReLU
1 1 Dropout
1 2 Linear

If this model is loaded, dropout layers are not used at all in the current approach. If it is used on a smaller dataset, it might easily become overfitted. In this case, having a dropout layer might be helpful.

[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:690]: @davidegraff Do you have any thoughts on this?

[the remark this answers is not in the document]
Comment (davidegraff) [reply chemprop/cli/train.py:690]: IMO dropout layers should be viewed in terms of what they're _preceding_ rather than what they're _succeeding_ when used as regularizers. So here, dropout is being used to regularize the inner layers by randomly masking out features in an attempt to suppress overfitting (i.e., decrease the layer's ability to rely on noise in the representation). The converse (viewing it as a regularizer of the preceding linear layer) doesn't really make sense to me because you can't zero out inputs of a layer that occurs before you.

File: chemprop/utils/v1_to_v2.py [hunk 2]
```diff
@@ -37,8 +37,11 @@ def convert_state_dict_v1_to_v2(model_v1_dict: dict) -> dict:
         ).unsqueeze(0)
 
     for i in range(args_v1.ffn_num_layers):
-        state_dict_v2[f"predictor.ffn.{i*3}.weight"] = state_dict_v1[f"readout.{i*3+1}.weight"]
-        state_dict_v2[f"predictor.ffn.{i*3}.bias"] = state_dict_v1[f"readout.{i*3+1}.bias"]
+        suffix = 0 if i == 0 else 2
+        state_dict_v2[f"predictor.ffn.{i}.{suffix}.weight"] = state_dict_v1[
+            f"readout.{i*3+1}.weight"
+        ]
+        state_dict_v2[f"predictor.ffn.{i}.{suffix}.bias"] = state_dict_v1[f"readout.{i*3+1}.bias"]
```
Comment (davidegraff) [inline chemprop/utils/v1_to_v2.py:44]:
```suggestion
        LINEAR_INDEX = -1
        state_dict_v2[f"predictor.ffn.{i}.{LINEAR_INDEX}.weight"] = state_dict_v1[
            f"readout.{i*3+1}.weight"
        ]
        state_dict_v2[f"predictor.ffn.{i}.{LINEAR_INDEX}.bias"] = state_dict_v1[f"readout.{i*3+1}.bias"]
```
the linear layer is always the "last" layer in a block (even for 1-layer blocks)

File: chemprop/cli/train.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (shihchengli) [reply chemprop/cli/train.py:690]: Thank you David for your insight. I appreciate the clarity on viewing dropout layers as regularizers for the layers that follow, rather than those that precede them. If I understand correctly, are we good to keep it as it is now?

**Comment by shihchengli** [comment]
@davidegraff, is this PR good to be merged now?

**Review by davidegraff — APPROVED** [review]
